Feature refactor survey answered ratio

idelcano opened the pull request:
Closes #1972

idelcano committed:
remove unnused imports

idelcano committed:
remove unused imports

idelcano committed:
Added survey question value mappers

idelcano committed:
update survey tests

idelcano committed:
remove unused import

idelcano committed:
Added question methods to control parent/children relations

idelcano committed:
create surveyAnsweredRatio constructor to use whitin survey

idelcano committed:
Added survey answer ratio in domain tests

idelcano committed:
update tests names

xurxodev reviewed (commented):
@idelcano this branch is based in an old state of the branch of the pull request  #2029, It's difficult to separate your changes of mine. I'll wait for the changes of #2029 to be mixed.

idelcano committed:
Merge branch 'development' into feature-refactor_survey_answered_ratio

idelcano committed:
Merge remote-tracking branch 'origin/feature-create_data_dhis_data_source' into feature-refactor_survey_answered_ratio

idelcano committed:
added compare without lists

xurxodev reviewed (changes requested):
@idelcano There are many conflicts to resolve, can you resolve it? I think is better option, thanks

idelcano committed:
Merge branch 'development' into feature-refactor_survey_answered_ratio

xurxodev committed:
Merge branch 'development' into feature-refactor_survey_answered_ratio

xurxodev commented on app/src/main/java/org/eyeseetea/malariacare/domain/entity/Survey.java:
```diff
@@ -18,35 +19,38 @@
 
     public static Survey createEmptySurvey(String uId, String programUId, String orgUnitUId,
-            String userUId) {
+            String userUId, HashMap<String, Question> questions) {
```
If It's easier to have a map inside of survey than a list, we should realize the conversion from list to map inside of survey because this conversion is realized one time, if we receive a map as an argument, all callers should convert from a list to map. Please change this to realize conversion inside the survey.

xurxodev commented on app/src/main/java/org/eyeseetea/malariacare/domain/entity/Survey.java:
```diff
@@ -123,48 +127,128 @@ public void changeScheduledDate(Date scheduledDate) {
+        return questions.get(uid);
+    }
+
+    public SurveyAnsweredRatio getSurveyAnsweredRatio() {
```
please rename to getAnsweredRatio, the word survey is redundant on this method because is a method of survey.

xurxodev commented on app/src/main/java/org/eyeseetea/malariacare/domain/entity/Survey.java:
```diff
@@ -123,48 +127,128 @@ public void changeScheduledDate(Date scheduledDate) {
+        return surveyAnsweredRatio;
+    }
+
+    public void setSurveyAnsweredRatio(SurveyAnsweredRatio surveyAnsweredRatio) {
```
this method is not necessary because is not called from outside, only is called inside of survey then we can use this.surveyAnsweredRatio directly

xurxodev commented on app/src/main/java/org/eyeseetea/malariacare/domain/entity/Survey.java:
```diff
@@ -123,48 +127,128 @@ public void changeScheduledDate(Date scheduledDate) {
+        this.surveyAnsweredRatio = surveyAnsweredRatio;
+    }
+
+    public void addValue(QuestionValue questionValue, Question question){
```
Question argument is not necessary because from questionValue argument we can get QustionUID and retrieve deseated question from questions existed questions map in survey

xurxodev commented on app/src/main/java/org/eyeseetea/malariacare/domain/entity/Survey.java:
```diff
@@ -123,48 +127,128 @@ public void changeScheduledDate(Date scheduledDate) {
+        getSurveyAnsweredRatio().addQuestion(question.isCompulsory());
+    }
+
+    public void removeValue(QuestionValue questionValue, Question question){
```
Question argument is not necessary because from questionValue argument we can get QustionUID and retrieve deseated question from questions existed questions map in survey

xurxodev commented on app/src/main/java/org/eyeseetea/malariacare/domain/entity/Question.java:
```diff
@@ -47,6 +67,67 @@ public String getAnswerName() {
 
+    public void addParent(Question question, String optionUid1) {
+        parents.add(question);
+        Map<String, Boolean> matches = new HashMap<>();
```
I don't understand well this code, please refactor this code to more semantic code, extract in methods if It's necessary where the names are explicative

xurxodev commented on app/src/main/java/org/eyeseetea/malariacare/domain/entity/Question.java:
```diff
@@ -47,6 +67,67 @@ public String getAnswerName() {
+    }
+
+    public void addActiveParentMatch(QuestionValue questionValue) {
+        if(questionValue.getQuestionUId()!=null && questionValue.getOptionUId()!=null){
```
I don't understand well this code, please refactor this code to more semantic code, extract in methods if It's necessary where the names are explicative

xurxodev commented on app/src/main/java/org/eyeseetea/malariacare/domain/entity/Question.java:
```diff
@@ -47,6 +67,67 @@ public String getAnswerName() {
+    }
+
+    public void removeActiveParentMatch(QuestionValue questionValue) {
+        if(questionValue.getQuestionUId()!=null && questionValue.getOptionUId()!=null){
```
I don't understand well this code, please refactor this code to more semantic code, extract in methods if It's necessary where the names are explicative

xurxodev commented on app/src/main/java/org/eyeseetea/malariacare/domain/entity/Question.java:
```diff
@@ -47,6 +67,67 @@ public String getAnswerName() {
+    }
+
+    private boolean hasActiveParent() {
+        for(String questionUId : questionOptionMatches.keySet()){
```
I don't understand well this code, please refactor this code to more semantic code, extract in methods if It's necessary where the names are explicative

xurxodev reviewed (changes requested):
@idelcano I have added some refactors to review as comments in the code

idelcano committed:
change HashMap param to List

idelcano committed:
refactor method names

idelcano committed:
remove unnecesary params

idelcano committed:
refactor addParent method to make it more readable

idelcano committed:
Refactor methods to make its more readable

idelcano commented:
@xurxodev done !

xurxodev committed:
Move methods to create maps from SurveyMapper to survey.
This SurveyMapper (SDK) is only to convert from event to map, adding this
methods here we added a responsibility more to mapper

xurxodev committed:
Refactor to more readable code

xurxodev commented on app/src/test/java/org/eyeseetea/malariacare/domain/entity/SurveyAnswerRatioShould.java:
```diff
@@ -0,0 +1,529 @@
+    public ExpectedException thrown = ExpectedException.none();
+
+    @Test
+    public void calculate_survey_ratio_when_add_normal_value(){
```
There are tests on this class that really are a survey tests, on SurveyAnswerRatioShould class should only test SurveyAnswerRatio class.  Please refactor tests moving it to SurveyShould class when necessary.

xurxodev commented on app/src/test/java/org/eyeseetea/malariacare/domain/entity/SurveyAnswerRatioShould.java:
```diff
@@ -0,0 +1,529 @@
+        Survey survey = createQuestionTreeWithoutMultiparentChildren();
+
+
+        Question question = survey.getQuestion("QuestionUID2");
```
there are similar code on several tests to create Question and QuestionValues, please refactor creating GivenA... methods to do the code more readable

xurxodev commented on app/src/main/java/org/eyeseetea/malariacare/domain/entity/Question.java:
```diff
@@ -1,25 +1,36 @@
-import static org.eyeseetea.malariacare.domain.utils.RequiredChecker.required;
+import java.util.Map;
 
 public class Question {
```
I still do not understand the logic that exists in question, can you create tests for question class that validate all this added logic that serves as documentation?

xurxodev reviewed (changes requested):
@idelcano We are slowly improving this part that is very important, I have requested for new changes

idelcano committed:
move tests

idelcano committed:
create tests

idelcano committed:
Merge branch 'development' into feature-refactor_survey_answered_ratio

idelcano committed:
Added answerratio tests

idelcano committed:
refactor surveyShould tests

idelcano committed:
remove imports

idelcano committed:
Added question tests and refactor survey

idelcano committed:
added new tests

idelcano committed:
Merge branch 'development' into feature-refactor_survey_answered_ratio

idelcano committed:
Merge branch 'development' into feature-refactor_survey_answered_ratio

idelcano committed:
fix survey percentaje

idelcano committed:
added test

idelcano committed:
fix test

idelcano committed:
update mockito to solve AbstractMethodError crash

xurxodev committed:
Merge branch 'development' into feature-refactor_survey_answered_ratio

xurxodev reviewed (changes requested):
@idelcano I see some drawbacks in this approach, such as having a very large graph and relationships between aggregates (root entities) sometimes but not always and I do not quite convince, I think the best thing is that we talk in several sessions pair programming to see this process. I need to understand the process very well, which is complex, to analyze the best way to represent it. We can talk about it in the next meeting with @ifoche since it is more priority to uncouple the pull and the new library than this.
